students/yixingxu/lesson06/mailroom4.py

```diff
@@ -0,0 +1,145 @@
+                # added exception handling in this version
+                try:
+                    add_donation_amount(name, float(donation_amount))
+                except ValueError:
+                    print('\nInvalid donation ammount. Please enter a valid number.')
+                else:
+                    break
+            print_thankyou_email(name,donation_amount)  
+        
+# print thank you email
+def print_thankyou_email(name,donation_amount):
+    print(f"\nDear {name},\n"
+        f"Thanks for your generous donation of ${donation_amount}"
+        )
+
+# construct list of names
+# use list comprehension 
+def list_name(donation_history = donation_history):
+    namelist = [item['name'] for item in donation_history]
+    return namelist     
+
+# add name to the list    
+def add_name(name,donation_history = donation_history):
+    donation_history.append({'name':name, 'donations':[]})
+    print('\nNew name {} has been added to the list'.format(name))
+    
+def add_donation_amount(name, donation_amount, donation_history = donation_history):
+    for item in donation_history:
+        if item['name'] == name:
+            item['donations'].append(donation_amount)
+            break
+    
+def create_a_report(donation_history=donation_history):
+    # create a summary list
+    summary_list = create_summary()
+    # sort summary list according to the total donation
+    sorted_list = summary_list[:]
+    sort_list(sorted_list)
+    print_report(sorted_list)
+    return sorted_list
+    
+# print a list
+def print_report(list):
+    donor_name_width = 20
+    total_given_width = 15
+    num_gifts_width = 5
+    average_gift_width = 15
+    seperator_width = 2
+    title = ['Donor Name','|','Total Given', '|', 'Num Gifts', '|', 'Average Gift']
+    divider = '-'*(donor_name_width+total_given_width+num_gifts_width+average_gift_width+seperator_width*5)
+    print('{:<20}{:<2}{:<15}{:<2}{:<5}{:<2}{:<15}'.format(*title))
+    print(divider)
+    for item in list:
+        print('{:<20}{:2}{:15.2f}  {:8} {:>2}{:12.2f}\n'.format(*item))
+     
+# create a summary list
+def create_summary(donation_history=donation_history):
+    summary_list = []
+    for item in donation_history:
+        total = sum(item['donations'])
+        number_of_gifts = len(item['donations'])
+        average = total/number_of_gifts
+        summary_list.append([item['name'],'$',total,number_of_gifts,'$',average])
+    return summary_list
+
+# sort summary list according to the total donation
+def sort_list(sorted_list):
+    #return summary_list.sort(key=sort_by_total_given)
+    sorted_list.sort(key=sort_by_total_given,reverse=True)
+
+# get the total given amount 
+def sort_by_total_given(list):
+    return list[2]
+    
+# quit
+def quit_program():
+    print('Quit')
+    return 'exit menu'
+
+def send_letters_to_everyone(donation_history = donation_history):
```

Comment:
this would be a good function to test

Reply:
can you give me a hint. How to test written out files? Thanks!

Reply:
in your test call `send_letters_to_everyone()`
then assert that files exists via `os.path.isfile`, in addition you can also open the file and assert contents of the file

Reply:
Thanks!